src/java.security.jgss/share/classes/org/ietf/jgss/GSSManager.java

```diff
@@ -423,6 +423,58 @@ public abstract GSSCredential createCredential (GSSName name,
                                   int lifetime, Oid mech, int usage)
         throws GSSException;
 
+    /**
+     * Factory method for acquiring a single mechanism credential with a
+     * password.<p>
+     *
+     * GSS-API mechanism providers must impose a local access-control
+     * policy on callers to prevent unauthorized callers from acquiring
+     * credentials to which they are not entitled. The kinds of permissions
+     * needed by different mechanism providers will be documented on a
+     * per-mechanism basis. A failed permission check might cause a {@link
+     * java.lang.SecurityException SecurityException} to be thrown from
+     * this method. <p>
+     *
+     * Non-default values for lifetime cannot always be honored by the
+     * underlying mechanisms, thus applications should be prepared to call
+     * {@link GSSCredential#getRemainingLifetime() getRemainingLifetime}
+     * on the returned credential.<p>
+     *
+     * @param name the name of the principal for whom this credential is to be
+     * acquired.  Use <code>null</code> to specify the default principal.
+     * @param lifetime The number of seconds that credentials should remain
+     * valid.  Use {@link GSSCredential#INDEFINITE_LIFETIME
+     * GSSCredential.INDEFINITE_LIFETIME} to request that the credentials
+     * have the maximum permitted lifetime.  Use {@link
+     * GSSCredential#DEFAULT_LIFETIME GSSCredential.DEFAULT_LIFETIME} to
+     * request default credential lifetime.
+     * @param mech the Oid of the desired mechanism.  Use <code>(Oid) null
+     * </code> to request the default mechanism.
+     * @param usage The intended usage for this credential object. The value
+     * of this parameter must be one of:
+     * {@link GSSCredential#INITIATE_AND_ACCEPT
+     * GSSCredential.INITIATE_AND_ACCEPT},
+     * {@link GSSCredential#ACCEPT_ONLY GSSCredential.ACCEPT_ONLY}, and
+     * {@link GSSCredential#INITIATE_ONLY GSSCredential.INITIATE_ONLY}.
+     * @return a GSSCredential of the requested type.
+     *
+     * @see GSSCredential
+     *
+     * @throws GSSException containing the following
+     * major error codes:
+     *    {@link GSSException#BAD_MECH GSSException.BAD_MECH},
+     *    {@link GSSException#BAD_NAMETYPE GSSException.BAD_NAMETYPE},
+     *    {@link GSSException#BAD_NAME GSSException.BAD_NAME},
+     *    {@link GSSException#CREDENTIALS_EXPIRED
+     *                                   GSSException.CREDENTIALS_EXPIRED},
+     *    {@link GSSException#NO_CRED GSSException.NO_CRED},
+     *    {@link GSSException#FAILURE GSSException.FAILURE}
+     */
+    public abstract GSSCredential createCredential (GSSName name,
+                                  String password, int lifetime, Oid mech,
```

Comment:
Shouldn't the password be a char array?

Reply:
Why should it be a `char` array and not a `String`?

Reply:
To securely wipe the password from memory against immutable strings in Java. In C you can safely to memset with zero, you Java you can't.

Reply:
Ahh, ok.

Reply:
Is that reliable?  @pburka suggests that a copying GC can easily leave copies of a password lying about.  And in the JAAS case the password would come from a file that is read in and parsed, so there will be a number of copies.  I'm not keen on reading the whole file as a `char[]` just so we could wipe the password.

Reply:
I think I'll let Oracle deal with this one.

Reply:
What file are you talking about? With JAAS you have callback handlers and they use char arrays too.

Reply:
Oh, maybe I was confused.  Krb5LoginModule let's you specify a raw secret key -- that's a bit crazy, IMO.  Still, other login modules use `String` for passwords too -- at least `JndiLoginModule`.

Reply:
How if I look at [this](https://docs.oracle.com/javase/7/docs/api/javax/security/auth/callback/PasswordCallback.html)?

Reply:
From `src/jdk.security.auth/share/classes/com/sun/security/auth/module/JndiLoginModule.java`

```
543                 // check the password
544                 if (verifyPassword
545                     (encryptedPassword, new String(password)) == true) {
...
710     /**
711      * Verify a password against the encrypted passwd from /etc/shadow
712      */
713     private boolean verifyPassword(String encryptedPassword, String password) {
...
```

Reply:
Ok, I see, that's ugly. You'll see a lot of such things in the JDK. Old Code (TM).